pkg/cmd/provisioning/alibabacloud/create-ram-users.go

```diff
@@ -0,0 +1,347 @@
+			//create new policy
+			log.Printf("Ready for creating new ram policy %s", policyName)
+			preq := ram.CreateCreatePolicyRequest()
+			preq.PolicyName = policyName
+			preq.PolicyDocument = string(policyBytes)
+			preq.Description = "Created by OpenShift ccoctl"
+			_, err = client.CreatePolicy(preq)
+			if err == nil {
+				return nil
+			}
+		}
+		log.Fatalf("Failed to create new ram policy %s, err is %v", policyName, err)
+	}
+	//create new policy with new version and set it to the default one
+	vreq := ram.CreateCreatePolicyVersionRequest()
+	vreq.Scheme = apiRequestScheme
+	vreq.PolicyName = policyName
+	vreq.PolicyDocument = string(policyBytes)
+	vreq.RotateStrategy = autoRotateStrategy
+	vreq.SetAsDefault = "true"
+	_, err = client.CreatePolicyVersion(vreq)
+	return err
+}
+
+//attachComponentPolicy will attach the given policy to the preset component user
+func attachComponentPolicy(client alibabacloud.Client, user, policyName string) error {
+	req := ram.CreateAttachPolicyToUserRequest()
+	req.Scheme = apiRequestScheme
+	req.PolicyName = policyName
+	req.UserName = user
+	req.PolicyType = ramPolicyType
+	_, err := client.AttachPolicyToUser(req)
+	return err
+}
+
+func processNewCredentialsRequests(client alibabacloud.Client, credReqs []*credreqv1.CredentialsRequest, name, targetDir string) error {
+	for _, cr := range credReqs {
+		err := createUserAndAttachPolicy(client, name, targetDir, cr)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func createUser(client alibabacloud.Client, name string, credReq *credreqv1.CredentialsRequest) (string, error) {
+	userName := fmt.Sprintf("%s-%s-%s", name, credReq.Namespace, credReq.Name)
+	shortName, displayName := generateRAMUserName(userName)
+
+	userReq := ram.CreateCreateUserRequest()
+	userReq.UserName = shortName
+	userReq.DisplayName = displayName
+
+	user, err := client.CreateUser(userReq)
+	if err != nil {
+		aErr, ok := err.(*alibabaerrors.ServerError)
+		if ok && aErr.ErrorCode() == errorUserAlreadyExists {
+			log.Printf("RAM User %s already exists, continuing", shortName)
+			return shortName, nil
+		} else {
+			return "", errors.Wrap(err, "failed to create RAM user")
+		}
+	}
+
+	log.Printf("Created RAM User: %s", user.User.UserName)
+	return user.User.UserName, nil
+}
+
+func generateUserAccessKeys(client alibabacloud.Client, userName string) (*ram.CreateAccessKeyResponse, error) {
+	accessKeyReq := ram.CreateCreateAccessKeyRequest()
+	accessKeyReq.UserName = userName
+	accessKeyResp, err := client.CreateAccessKey(accessKeyReq)
+	if err != nil {
+		aErr, ok := err.(*alibabaerrors.ServerError)
+		if ok && aErr.ErrorCode() == errorAKLimitExceeded {
+			log.Printf("RAM User %s's accesskey number limit exceeded, will delete original accesskeys", userName)
+			listKeyReq := ram.CreateListAccessKeysRequest()
+			listKeyReq.UserName = userName
+			listKeyRes, err := client.ListAccessKeys(listKeyReq)
+			for _, oneAK := range listKeyRes.AccessKeys.AccessKey {
```

Comment:
Rather than delete all access keys that already exist for the user, can we leave the access key with the most recent "Created Date", and only clear out the older ones?

Reply:
has refine the code to choose the older one or the ones with in-active status